hris_publisher/hris.py

```diff
@@ -0,0 +1,187 @@
+        self.bucket_name = None
+
+    def load(self):
+        """Fetch the file as stream from the s3 bucket."""
+        if self.from_file is True:
+            hris_json = os.path.join(
+                os.path.abspath(os.path.dirname(__file__)),
+                '../sample_data/{file_name}'.format(file_name=self.file_name)
+            )
+
+            with open(hris_json, 'r') as schema_data:
+                hris_json = json.load(schema_data)
+        else:
+            # Go get it from s3 using the helper functions
+            self._locate_s3_bucket()
+            streaming_body = self._get_object_from_bucket()
+            hris_json = self._read_object(streaming_body)
+        return hris_json
+
+    def is_valid(self, entry):
+        """
+        Validate the object against the HRIS schema json file included in the project.
+        :param entry: a single json entry from the hris.json
+        :return: bool (truthy)
+        """
+        try:
+            hris_schema_json = os.path.join(os.path.abspath(os.path.dirname(__file__)), 'schema.json')
+
+            with open(hris_schema_json, 'r') as schema_data:
+                hris_schema_json = json.load(schema_data)
+
+            jsonschema_validate(entry, hris_schema_json)
+            return True
+        except ValidationError:
+            return False
+
+    def to_groups(self, entry):
+        """
+
+        :param entry: from the hris.json
+        :return: a list of groups to reintegrate to cis beginning with hris_
+        """
+        g = Groups(entry=entry)
+        return g.all
+
+    def _locate_s3_bucket(self):
+        """Fine the S3 bucket in the account using bucket tags."""
+        if self.boto_session is None:
+            self.boto_session = self._connect_boto_session()
+
+        client = self.boto_session.client('s3', region_name='us-west-2')
+        s3 = self.boto_session.resource('s3')
+        for this_bucket in client.list_buckets().get('Buckets'):
+            bucket_name = this_bucket.get('Name')
+            bucket_tagging = s3.BucketTagging(bucket_name)
+
+            try:
+                tag_set = bucket_tagging.tag_set
+                if self._is_tag_match(tag_set):
+                    self.bucket_name = bucket_name
+                    return bucket_name
+            except ClientError:
+                # Boto3 throws client error when a bucket has no tags.
+                continue
+
+    def _is_tag_match(self, tag_set):
+        for tag in tag_set:
+            if tag.get('Value') == 'hris_publisher':
+                return True
+        return False
+
+    def _connect_boto_session(self):
+        return boto3.session.Session()
+
+    def _get_object_from_bucket(self):
+        """Get the byte stream of the hris.json file."""
+        if self.boto_session is None:
+            self.boto_session = self._connect_boto_session()
+
+        client = self.boto_session.client('s3', region_name='us-west-2')
+        response = client.get_object(
```

Comment:
do we want to error out cleanly or log the error if `self.bucket_name` is None? (ie no bucket was tagged with our tag)

Reply:
irc chat: prefer traceback and exit on all conditions

Reply:
The custom exception for no bucket found is fair.  I will do that one.